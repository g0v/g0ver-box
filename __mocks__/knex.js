import knex from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default knex;

process.setMaxListeners(0);
