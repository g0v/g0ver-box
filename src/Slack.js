import _ from 'lodash';
import slack from 'slack';

export const bot = slack.rtm.client();

const TOKEN = process.env.SLACK_TOKEN;

if (TOKEN) {
  bot.listen({ token: TOKEN });
}

bot.hello((message) => {
  console.log(`Got a message: ${message.type}`);
});

const call = (name, req = {}) => (
  new Promise((resolve, reject) => {
    const fn = _.get(slack, name);
    fn(
      { ...req, token: TOKEN },
      (err, res) => (err ? reject(err) : resolve(res))
    );
  })
);

export default {
  userInfo: async req => await call('users.info', req),
  postMessage: async req => await call('chat.postMessage', { ...req, as_user: true }),
  postMultiMessage: async (users, message) => _.map(users, user => bot.ws.send(JSON.stringify({
    id: `${user}::${Date.now()}`, type: 'message', channel: user, text: message,
  }))),
  channelInfo: async req => await call('channels.info', req),
  channelJoin: async req => await call('channels.join', req),
};
