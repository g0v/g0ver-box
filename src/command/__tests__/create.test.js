import { client } from 'knex';
import Slack from '../../Slack';
import index from '../';

describe('create command', () => {
  const data = { channel: 'D100', user: 'U03B2AB13', name: 'yutin', text: 'create g0ver-box' };

  it('when project does not exist', async () => {
    client.mockReturnValueOnce([]);
    client.mockReturnValueOnce(['5']);

    await index(data);
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: '請輸入 共筆 / 討論 / 專業 連結？\n`（輸入 skip 可跳過）`',
    });

    await index({ ...data, text: 'https://project.g0v.tw/' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: '請輸入專案的 關鍵字 ，多個關鍵字請用空白分隔 ？\n`（輸入 skip 可跳過）`',
    });

    await index({ ...data, text: '找人 找坑' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: '請輸入專案的縮圖網址？支援 GIF, JPEG, PNG 格式的 75px X 75px\n`（輸入 skip 可跳過）`',
    });

    await index({ ...data, text: 'https://project.g0v.tw/index.png' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: 'Done, create g0ver-box',
    });

    expect(client).toMatchSnapshot();
  });

  it('when project is existed', async () => {
    client.mockReturnValueOnce([{ id: 5 }]);

    await index(data);
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: 'Sorry! 發現相同的專案名稱',
    });

    expect(client).toMatchSnapshot();
  });

  it('when use skip', async () => {
    client.mockReturnValueOnce([]);
    client.mockReturnValueOnce(['5']);

    await index(data);
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: '請輸入 共筆 / 討論 / 專業 連結？\n`（輸入 skip 可跳過）`',
    });

    await index({ ...data, text: 'skip' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: '請輸入專案的 關鍵字 ，多個關鍵字請用空白分隔 ？\n`（輸入 skip 可跳過）`',
    });

    await index({ ...data, text: 'skip' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: '請輸入專案的縮圖網址？支援 GIF, JPEG, PNG 格式的 75px X 75px\n`（輸入 skip 可跳過）`',
    });

    await index({ ...data, text: 'skip' });
    expect(Slack.postMessage).toHaveBeenLastCalledWith({
      channel: 'D100',
      text: 'Done, create g0ver-box',
    });

    expect(client).toMatchSnapshot();
  });
});
