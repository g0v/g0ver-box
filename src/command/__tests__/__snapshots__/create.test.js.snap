// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create command when project does not exist 1`] = `
Array [
  "select * from project where deleted_at is null and title = g0ver-box limit 1",
  "insert into project (archive, created_at, keyword, title, updated_at, user_id) values ({\\"url\\":\\"https://project.g0v.tw/\\",\\"tags\\":\\"找人 找坑\\",\\"thumb\\":\\"https://project.g0v.tw/index.png\\"}, DATE, 找人 找坑, g0ver-box, DATE, U03B2AB13) returning id",
]
`;

exports[`create command when project is existed 1`] = `
Array [
  "select * from project where deleted_at is null and title = g0ver-box limit 1",
]
`;

exports[`create command when use skip 1`] = `
Array [
  "select * from project where deleted_at is null and title = g0ver-box limit 1",
  "insert into project (archive, created_at, keyword, title, updated_at, user_id) values ({\\"url\\":null,\\"tags\\":null,\\"thumb\\":null}, DATE, , g0ver-box, DATE, U03B2AB13) returning id",
]
`;
